target/bin/updatemailuser

```diff
@@ -10,25 +10,79 @@
 
 DATABASE=${DATABASE:-/tmp/docker-mailserver/postfix-accounts.cf}
 
-USER="${1}"
+MAIL_CRYPT_ENABLED=false
+
+function __usage
+{
+  echo -e "\e[35mUPDATEMAILUSER\e[31m(\e[93m8\e[31m)
+
+\e[38;5;214mNAME\e[39m
+    updatemailuser - allows user to update an existing email user
+      If mail_crypt is enabled, you'll need to provide the email user's old password.
+
+\e[38;5;214mSYNOPSIS\e[39m
+    updatemailuser <EMAIL ADDRESS> [<PASSWORD>] [<OLD CRYPTOKEY PASSWORD>\e[31m...\e[39m] \e[31m|\e[39m help]
+
+\e[38;5;214mOPTIONS\e[39m
+    \e[94mGeneric Program Information\e[39m
+        help       Print the usage information.
+
+\e[38;5;214mEXIT STATUS\e[39m
+    Exit status is 0 if command was successful, and 1 if there was an error.
+"
+}
+
+if [[ ${1} == 'help' ]]
+then
+  __usage
+  exit 0
+fi
+
+# Check if mail_crypt/per user storage encryption is enabled
+if ! grep '^#.*auth-passwdfile-mailcrypt' /etc/dovecot/conf.d/10-auth.conf &>/dev/null
+then
+  MAIL_CRYPT_ENABLED=true
+fi
+
+FULL_EMAIL="${1}"
 shift
 PASSWD="${*}"
+shift
+CRYPTOKEY_UPDATE_PASSWORD="${*}"
 
-function usage  {	echo "Usage: updatemailuser <user@domain.tld> [password]" ; }
+[[ -z ${FULL_EMAIL} ]] && { usage ; errex "no username specified" ; }
 
-[[ -z ${USER} ]] && { usage ; errex "no username specified" ; }
+USER="${FULL_EMAIL%@*}"
+DOMAIN="${FULL_EMAIL#*@}"
 
 if [[ -z ${PASSWD} ]]
 then
-  read -r -s -p "Enter Password: " PASSWD
+  read -r -s -p "Enter New Password: " PASSWD
   echo
-  [[ -z ${PASSWD} ]] && errex "Password must not be empty"
+  [[ -z ${PASSWD} ]] && errex "password must not be empty"
 fi
 
-HASH="$(doveadm pw -s SHA512-CRYPT -u "${USER}" -p "${PASSWD}")"
+HASH="$(doveadm pw -s SHA512-CRYPT -u "${FULL_EMAIL}" -p "${PASSWD}")"
```

Comment:
Do you know if there will be any compatibility problems if we ever change from `SHA512-CRYPT`, especially with existing mail server setups?

I was looking into this a few months ago but got side tracked, IIRC it wouldn't be an issue. I was investigating it because of earlier discussions on the API for password changing where I think there was some assumption of SHA512-CRYPT and talk of potentially using that assumption in the API logic outside of this project.

I mention it since some newer distro releases are switching user account passwords (unrelated to dovecot) from SHA512-CRYPT to yescrypt. SHA512-CRYPT has some minor security disadvantages but getting bcrypt and similar alternatives was often blocked by upstream who implemented SHA512-CRYPT. With our current debian base packages, we do have access to bcrypt and argon2id IIRC.

Reply:
If you'd like me to still give a review I should be able to within the next day or so, let me know :)

Reply:
Maybe a side convo should be opened about this. Correct me if I'm wrong but this doesn't seem related to this feature since it has always been set to SHA512-CRYPT?

Reply:
> Correct me if I'm wrong but this doesn't seem related to this feature

Yeah not related, just asking since you've been touching code around it and involved with the API project, as it may be a design concern to keep in mind.

I'm not requesting you change anything here, just that it may change in future (either by us or upstream) at some point and if you see that being fragile with what you're doing here or the API project.

Reply:
Understood -- thanks for clarifying! I'll think about it a bit but it might be a bit over my head :(